core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp

```diff
@@ -523,44 +523,55 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
     int max_active_league_size =
         OpenMPTargetExec::MAX_ACTIVE_THREADS / team_size;
 
-    // A lock array to keep track of scratch memory blocks.
-    // The first element to keep track of the number of free-blocks.
-    int* lock_array = new int[max_active_league_size + 1];
-    memset(lock_array, 1, (max_active_league_size + 1) * sizeof(int));
-    lock_array[0] = max_active_league_size;
+    // Allocate the lock_array on the device and initialize it to 0's in a
+    // target region.
+    int* lock_array = static_cast<int*>(omp_target_alloc(
+        (max_active_league_size + 1) * sizeof(int), omp_get_default_device()));
+
+#pragma omp target teams distribute parallel for is_device_ptr(lock_array)
+    for (int i = 1; i < max_active_league_size + 1; ++i) {
+      lock_array[i] = 0;
+
+      // Only one thread in one team sets the number of free blocks as the value
+      // of the first index in the lock_array.
+      if (omp_get_team_num() == 0 && omp_get_thread_num() == 0)
+        lock_array[0] = max_active_league_size;
+    }
 
-#pragma omp target teams distribute map(                    \
-    to                                                      \
-    : a_functor, lock_array [0:max_active_league_size + 1]) \
-    is_device_ptr(scratch_ptr) num_teams(nteams) thread_limit(team_size)
+#pragma omp target teams distribute map(to                   \
+                                        : a_functor)         \
+    is_device_ptr(scratch_ptr, lock_array) num_teams(nteams) \
+        thread_limit(team_size)
     for (int i = 0; i < league_size; i++) {
-      int shmem_block_index = 0, lock_team = 99999;
-
-      // Allocate the first set of max_active_league_size blocks.
-      if (i < max_active_league_size) {
-        shmem_block_index = i + 1;
-        atomic_decrement(&lock_array[0]);
-        lock_array[shmem_block_index] = 0;
-      } else {
-        // Block until there is atleast one free scratch block is available.
-        while (lock_array[0] == 0)
-          ;
-
-        // Iterate till the first free scratch block is encountered and try and
-        // get a lock on the block.
-        int iter = 1;
-        while (lock_array[iter] != 1) iter++;
-        shmem_block_index = iter;
-
-        lock_team =
-            atomic_compare_exchange(&lock_array[shmem_block_index], 1, 0);
-        while (lock_array[shmem_block_index] == 1) {
-          lock_team =
-              atomic_compare_exchange(&lock_array[shmem_block_index], 1, 0);
+      int shmem_block_index = 0, lock_team = 99999, iter = 0;
+
+      // Loop as long as a shmem_block_index is not found.
+      while (shmem_block_index == 0) {
+        // Proceed if there are free blocks available.
+        if (lock_array[0] > 0) {
```

Comment:
I don't think this is needed. This should ALWAYS be true. The point of the lock array is that it is actually a bit larger then the max number of active teams. So the atomic counter can never reach zero ...